tests/conftest.py

```diff
@@ -14,7 +14,140 @@
 
+from collections.abc import AsyncGenerator, Generator
+
+import pytest
+import pytest_asyncio
+from hexkit.custom_types import PytestScope
 from hexkit.providers.mongodb.testutils import (  # noqa: F401
     mongodb_container_fixture,
     mongodb_fixture,
 )
+from hexkit.providers.s3 import S3ObjectStorage
+from hexkit.providers.s3.testutils import (
+    S3ContainerFixture,
+    S3Fixture,
+    temp_file_object,
+)
+
+from sms.config import Config
+from tests.fixtures.config import get_config
+
+
+# Move the federation fixtures to hexkit if deemed useful
+class FederatedS3Fixture:
+    """Fixture containing multiple S3 fixtures to simulate federated storage."""
+
+    def __init__(self, storages: dict[str, S3Fixture]):
+        self.nodes = storages
+
+    def get_patched_config(self, *, config: Config):
+        """Update a Config instance so the object storage credentials point to the
+        respective S3Fixtures.
+        """
+        config_vals = config.model_dump()
+        for alias in config.object_storages:
+            node_config = self.nodes[alias].config.model_dump()
+            config_vals["object_storages"][alias]["credentials"] = node_config
+        return Config(**config_vals)
+
+    async def populate_dummy_items(self, alias: str, buckets: dict[str, list[str]]):
+        """Convenience function to populate a specific S3Fixture.
+
+        The `buckets` arg is a dictionary with bucket names as keys and lists of objects
+        as values. The buckets can be empty, and the objects are created with a size of
+        1 byte. This function might be modified or removed in the hexkit version.
+        """
+        if alias not in self.nodes:
+            # This would indicate some kind of mismatch between config and fixture
+            raise RuntimeError(f"Alias '{alias}' not found in the federated S3 fixture")
+        storage = self.nodes[alias]
+        # Populate the buckets so even empty buckets are established
+        await storage.populate_buckets([bucket for bucket in buckets])
+        for bucket, objects in buckets.items():
+            for object in objects:
+                with temp_file_object(bucket, object, 1) as file:
+                    await storage.populate_file_objects([file])
+
+
+class MultiS3ContainerFixture:
+    """Fixture for managing multiple running S3 test containers in order to mimic
+    multiple object storages.
+
+    Without this fixture, separate S3Fixture instances would access the same
+    underlying storage resources.
+    """
+
+    def __init__(self, s3_containers: dict[str, S3ContainerFixture]):
+        self.s3_containers = s3_containers
+
+        for container in self.s3_containers.values():
+            container.__enter__()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Exit the context manager and clean up the S3 containers."""
+        for container in self.s3_containers.values():
+            container.__exit__(exc_type, exc_val, exc_tb)
```

Comment:
A bit confused about the missing symmetry (i.e. no `__enter__`), but I guess `__exit__` is defined due to some exception shenanigans during testing?

Reply:
No you're right, I guess I was in between two iterations and didn't catch it because the tests ran. It should be fixed in [7e478fc](https://github.com/ghga-de/state-management-service/pull/5/commits/7e478fc7e7f2c216416ffdb5fa08e57ad56bacc4)